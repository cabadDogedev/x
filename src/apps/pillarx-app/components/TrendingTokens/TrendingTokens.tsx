import { useTranslation } from 'react-i18next'
import { createRef, useEffect, useState } from 'react';

// types
import { TrendingTokenData } from '../../../../types/api'

// hooks
import useRefDimensions from '../../hooks/useRefDimensions';

// images

// components
import TileContainer from '../TileContainer/TileContainer'
import Body from '../Typography/Body';
import TrendingTokenInfo from '../TrendingTokenInfo/TrendingTokenInfo';
import SkeletonLoader from '../../../../components/SkeletonLoader';


type TrendingTokensProps = {
    data: TrendingTokenData[] | undefined;
    isDataLoading: boolean;
}

const TrendingTokens = ({ data, isDataLoading }: TrendingTokensProps) => {
    const [t] = useTranslation();

    const [trendingTokenWidth, setTrendingTokenWidth] = useState<number>(0);

    useEffect(() => {
        const handleTrendingTokenWidthResize = () => {
            if (window.innerWidth >= 1024) {
                setTrendingTokenWidth(159);
            } else if (window.innerWidth >= 800) {
                setTrendingTokenWidth(152);
            } else {
                // 102px for width of the TrendingTokenInfo component 100px + 2px of space
                setTrendingTokenWidth(102);
            }
        };

        handleTrendingTokenWidthResize();
        window.addEventListener('resize', handleTrendingTokenWidthResize);

        return () => {
            window.removeEventListener('resize', handleTrendingTokenWidthResize);
        };
    }, []);
    
    const divRef = createRef()
    const dimensions = useRefDimensions(divRef as React.RefObject<HTMLDivElement>)

    const numberTrendingTokens = Math.floor(dimensions.width / trendingTokenWidth) ?? 0;


if (!data || isDataLoading) {
    return (
    <TileContainer className='flex-col px-10 pt-[30px] pb-5 tablet:p-5 mobile:p-0 mobile:bg-[#1F1D23]'>
         <SkeletonLoader  $height='24px' $width='140px' $marginBottom='10px' $radius='10px' />
         <div className='flex justify-between gap-2'>
         <SkeletonLoader  $height='120px' $width='80px' $radius='10px' />
         <SkeletonLoader  $height='120px' $width='80px' $radius='10px' />
         <SkeletonLoader  $height='120px' $width='80px' $radius='10px' />
         <SkeletonLoader  $height='120px' $width='80px' $radius='10px' className='tablet:hidden' />
         <SkeletonLoader  $height='120px' $width='80px' $radius='10px' className='tablet:hidden' />
         <SkeletonLoader  $height='120px' $width='80px' $radius='10px' className='tablet:hidden' />
         <SkeletonLoader  $height='120px' $width='80px' $radius='10px' className='mobile:hidden' />
         <SkeletonLoader  $height='120px' $width='80px' $radius='10px' className='mobile:hidden' />
         </div>
    </TileContainer>
    )
}

    return (
        <div ref={divRef as React.RefObject<HTMLDivElement>}>
            <TileContainer className='flex-col px-10 pt-[30px] pb-5 tablet:p-5 mobile:p-0 mobile:bg-[#1F1D23]'>
                <Body className='text-purple_light mb-2.5'>{t`Trending tokens`}</Body>
                <div className='flex justify-between'>
                {data?.slice(0, numberTrendingTokens).map((token, index) =>
                    <TrendingTokenInfo key={index} logo={token.logo} tokenName={token.name} tokenValue={undefined} percentage={undefined}  />
                )}
                </div>
            </TileContainer>
        </div>
    )
}

export default TrendingTokens;
