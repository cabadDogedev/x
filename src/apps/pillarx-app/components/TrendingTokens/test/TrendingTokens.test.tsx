import renderer from 'react-test-renderer';

// components
import TrendingTokens from '../TrendingTokens';

describe('<TrendingTokens />', () => {
    const mockData = [{
        id: 1,
        name: 'The best token',
        symbol: 'TBT',
        contracts: [{
            address: '0x123',
            blockchain: '111222',
            decimals: 18,
        }],
        logo: 'https://www.logo.com/example/',
        trending_score: 20,
        platforms: [{
            name: 'platform one',
            rank: 3,
            weight: 1,
    }],
        price_change_24h: 12.56,
        pair: 'pair example',
    }];

    const mockLoading = false;
  
    it('renders correctly and matches snapshot', () => {
        const tree = renderer
            .create(<TrendingTokens data={mockData} isDataLoading={mockLoading} />)
            .toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('displays loading skeleton when data is loading', () => {
        const tree = renderer
            .create(<TrendingTokens data={undefined} isDataLoading={true} />)
            .toJSON();

        expect(tree).toMatchSnapshot();
    });

});
