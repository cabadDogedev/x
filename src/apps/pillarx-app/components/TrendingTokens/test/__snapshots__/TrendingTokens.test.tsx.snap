// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TrendingTokens /> displays loading skeleton when data is loading 1`] = `
.c0 {
  height: 24px;
  width: 140px;
  animation: hLhBjP 1s linear infinite alternate;
  border-radius: 10px;
  margin-bottom: 10px;
}

.c1 {
  height: 120px;
  width: 80px;
  animation: hLhBjP 1s linear infinite alternate;
  border-radius: 10px;
  margin-bottom: 0;
}

<div
  className="flex bg-container_grey rounded-2xl flex-col px-10 pt-[30px] pb-5 tablet:p-5 mobile:p-0 mobile:bg-[#1F1D23]"
>
  <div
    className="c0"
  />
  <div
    className="flex justify-between gap-2"
  >
    <div
      className="c1"
    />
    <div
      className="c1"
    />
    <div
      className="c1"
    />
    <div
      className="c1 tablet:hidden"
    />
    <div
      className="c1 tablet:hidden"
    />
    <div
      className="c1 tablet:hidden"
    />
    <div
      className="c1 mobile:hidden"
    />
    <div
      className="c1 mobile:hidden"
    />
  </div>
</div>
`;

exports[`<TrendingTokens /> renders correctly and matches snapshot 1`] = `
<div>
  <div
    className="flex bg-container_grey rounded-2xl flex-col px-10 pt-[30px] pb-5 tablet:p-5 mobile:p-0 mobile:bg-[#1F1D23]"
  >
    <p
      className="text-base font-medium text-purple_light mb-2.5"
    >
      Trending tokens
    </p>
    <div
      className="flex justify-between"
    >
      <div
        className="flex flex-col py-5 px-[22px] gap-1 w-[122px] h-auto items-center tablet:w-[120px] mobile:w-[100px] mobile:px-3.5"
      >
        <img
          className="w-[70px] h-[70px] object-fill rounded-full mb-3.5"
          src="https://www.logo.com/example/"
        />
        <p
          className="text-base font-medium text-center"
        >
          The best token
        </p>
      </div>
    </div>
  </div>
</div>
`;
